python/ray/tune/schedulers/pb2.py

```diff
@@ -0,0 +1,322 @@
+                 quantile_fraction: float = 0.25,
+                 log_config: bool = True,
+                 require_attrs: bool = True,
+                 synch: bool = False):
+
+        resample_probability = 0
+        custom_explore_fn = None
+
+        super(PopulationBasedTraining, self).__init__(
+            self, time_attr, reward_attr, metric, mode, perturbation_interval,
+            hyperparam_mutations, quantile_fraction, resample_probability,
+            custom_explore_fn, log_config, require_attrs, synch)
+
+        self._name = "pb2"
+
+        self.latest = 0  # when we last explored
+        self.data = pd.DataFrame()
+
+        self.bounds = {}
+        for key, distribution in self._hyperparam_mutations.items():
+            self.bounds[key] = [
+                np.min([distribution() for _ in range(999999)]),
+                np.max([distribution() for _ in range(999999)])
+            ]
+
+    def _save_trial_state(self, state, time, result, trial):
+        score = super(PopulationBasedTraining, self)._save_trial_state(
+            state, time, result, trial)
+
+        # Data logging for PB2.
+
+        # Collect hyperparams names and current values for this trial.
+        names = []
+        values = []
+        for key in self._hyperparam_mutations.keys():
+            names.append(str(key))
+            values.append(trial.config[key])
+
+        # Store trial state and hyperparams in dataframe.
+        # this needs to be made more general.
+        lst = [[trial, result[self._time_attr]] + values + [score]]
+        cols = ['Trial', 'T'] + names + ['Reward']
+        entry = pd.DataFrame(lst, columns=cols)
+
+        self.data = pd.concat([self.data, entry]).reset_index(drop=True)
+        self.data.Trial = self.data.Trial.astype('str')
+
+    def _get_new_config(self, trial, trial_to_clone):
+        # If we are at a new timestep, we dont want to penalise for trials
+        # still going
+        if self.data['T'].max() > self.latest:
+            self.current = None
+
+        new_config, data = explore(self.data, self.bounds, self.current,
+                                   trial_to_clone, trial,
+                                   trial_to_clone.config)
+
+        # important to replace the old values, since we are copying across
+        self.data = data.copy()
+
+        # if the current guy youre selecting is at a point youve already done,
+        # then append the data to the "current" which is the points in the current batch
+
+        new = []
+        for key in self._hyperparam_mutations.keys():
+            new.append(new_config[key])
+
+        new = np.array(new)
+        new = new.reshape(1, new.size)
+        if self.data['T'].max() > self.latest:
+            self.latest = self.data['T'].max()
+            self.current = new.copy()
+        else:
+            self.current = np.concatenate((self.current, new), axis=0)
+            print(self.current)
```

Comment:
Can you replace this to use logging.debug instead and also add some description of what's being logged. Or if you don't think this is important for debugging then we can just remove this.

Reply:
Sorry I mean logger.debug here!

Reply:
Done.